src/mlpack/methods/ne/neat.hpp

```diff
@@ -0,0 +1,974 @@
+        bool hasBaby = BreedChild(aPopulation.aSpecies[i], genome, aCrossoverRate);
+        if (hasBaby) {
+          childGenomes.push_back(genome);
+          ++numBreedSuccess;
+        }
+      }
+      ////printf("hehe 3-2\n"); // DEBUG
+    }
+    ////printf("hehe 4\n"); // DEBUG
+
+    // Keep the best in each species.
+    //DEBUGGING!!!!!!!!!
+    //printf("before cull to one, species size is: %d \n", aPopulation.aSpecies.size());
+    printf("before, species sizes are: ");
+    for (ssize_t s=0; s<aPopulation.aSpecies.size(); ++s) {
+      std::cout<< aPopulation.aSpecies[s].aGenomes.size() << "  ";
+    }
+    printf("\n");
+    //DEBUGGING!!!!!!!!!
+    CullSpeciesToOne(aPopulation);
+    ////printf("hehe 5\n"); // DEBUG
+    //DEBUGGING!!!!!!!!!
+    //printf("after cull to one, species size is: %d \n", aPopulation.aSpecies.size());
+    printf("after, species sizes are: ");
+    for (ssize_t s=0; s<aPopulation.aSpecies.size(); ++s) {
+      std::cout<< aPopulation.aSpecies[s].aGenomes.size() << "  ";
+    }
+    printf("\n");
+    //DEBUGGING!!!!!!!!!
+
+    // Random choose species and breed child until reach population size.
+    childGenomes.push_back(lastBestGenome);
+    while (childGenomes.size() + aPopulation.aSpecies.size() < aPopulationSize) {
+      ssize_t speciesIndex = mlpack::math::RandInt(0, aPopulation.aSpecies.size());
+      Genome genome;  //!!!!!!!!!!!!!
+      bool hasBaby = BreedChild(aPopulation.aSpecies[speciesIndex], genome, aCrossoverRate);
+      if (hasBaby) childGenomes.push_back(genome);
+      ////printf("childGenomes size is %d \n", childGenomes.size());
+      ////printf("aSpecies size is %d \n", aPopulation.aSpecies.size());
+      ////printf("aPopulationSize is %d \n", aPopulationSize);
+    }
+    ////printf("hehe 6\n"); // DEBUG
+
+    // Speciate genomes into new species.
+    std::vector<Genome> populationGenomes;
+    AggregateGenomes(aPopulation, populationGenomes);
+    aPopulation.aSpecies.clear();
+    for (ssize_t i=0; i<populationGenomes.size(); ++i) {
+      AddGenomeToSpecies(aPopulation, populationGenomes[i]);
+    }
+    for (ssize_t i=0; i<childGenomes.size(); ++i) {
+      AddGenomeToSpecies(aPopulation, childGenomes[i]);
+    }
+    ////printf("hehe 7\n"); // DEBUG
+
+    // Reassign genome IDs.
+    aPopulation.ReassignGenomeId();
+  }
+
+  // Evaluate genomes in population, set genomes' fitness.
+  void Evaluate() {
+    for (ssize_t i=0; i<aPopulation.aSpecies.size(); ++i) {
+      for (ssize_t j=0; j<aPopulation.aSpecies[i].aGenomes.size(); ++j) {
+        ////printf("start eval fitness\n");
+        double fitness = aTask.EvalFitness(aPopulation.aSpecies[i].aGenomes[j]);
+       ////printf("end eval fitness\n");
+        aPopulation.aSpecies[i].aGenomes[j].Fitness(fitness);
+      }
+
+      double oldSpeciesBestFitness = aPopulation.aSpecies[i].BestFitness();
+      aPopulation.aSpecies[i].SetBestFitnessAndGenome();
```

Comment:
The SetBestFitnessAndGenome returns the minimal fitness over all genomes, but for the CartPole task as it implemented right now a higher fitniss is better.